nautobot/ipam/models.py

```diff
@@ -836,29 +835,42 @@ def get_first_available_ip(self):
         return f"{next(available_ips.__iter__())}/{self.prefix_length}"
 
     def get_utilization(self):
-        """Get the child prefix size and parent size.
+        """Return the utilization of this prefix as a UtilizationData object.
 
-        For Prefixes with a type of "container", get the number child prefixes. For all others, count child IP addresses.
+        For prefixes containing other prefixes, all direct child prefixes are considered fully utilized.
+
+        For prefixes containing IP addresses and/or pools, pools are considered fully utilized while
+        only IP addresses that are not contained within pools are added to the utilization.
 
         Returns:
             UtilizationData (namedtuple): (numerator, denominator)
         """
-        if self.type == choices.PrefixTypeChoices.TYPE_CONTAINER:
-            child_prefixes = netaddr.IPSet(p.prefix for p in self.descendants())
-            return UtilizationData(numerator=child_prefixes.size, denominator=self.prefix.size)
+        denominator = self.prefix.size
 
+        if self.type == choices.PrefixTypeChoices.TYPE_POOL:
+            pool_ips = IPAddress.objects.filter(
+                parent__namespace=self.namespace, host__gte=self.network, host__lte=self.broadcast
+            ).values_list("host", flat=True)
+            child_ips = netaddr.IPSet(pool_ips)
         else:
-            prefix_size = self.prefix.size
-            if all(
-                [
-                    self.ip_version == 4,
-                    self.prefix_length < 31,
-                    self.type != choices.PrefixTypeChoices.TYPE_POOL,
-                ]
-            ):
-                prefix_size -= 2
-            child_count = prefix_size - self.get_available_ips().size
-            return UtilizationData(numerator=child_count, denominator=prefix_size)
+            child_ips = netaddr.IPSet(self.ip_addresses.values_list("host", flat=True))
+
+        child_prefixes = netaddr.IPSet(p.prefix for p in self.children.only("network", "prefix_length").iterator())
+        numerator = child_prefixes | child_ips
+
+        # Exclude network and broadcast address from the denominator unless they've been assigned to an IPAddress or child pool.
+        if all(
+            [
+                not self.children.exclude(type=choices.PrefixTypeChoices.TYPE_POOL).exists(),
+                denominator > 2,
+                self.type != choices.PrefixTypeChoices.TYPE_POOL,
+                self.ip_version == 4,
```

Comment:
I saw your comment on Slack about the behavior of `all()`, but can we make this any more efficient by doing the non-DB-accessing checks first? I.e not even look at `self.children` if we don't pass the other three checks?